org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -347,8 +347,17 @@ private void createBufferFD() throws SSLException {
             throw new SSLException("Unable to enable SSL Handshake Callback on this SSLFDProxy instance.");
         }
 
-        // Pass this ssl_fd to the session object so that we can use
-        // SSL methods to invalidate the session.
+        if (alpn_protocols != null) {
```

Comment:
should also check `alpn_protocols.length > 0`.

Reply:
What `SSLParameters` class? We explicitly downcast into `JSSParameters` [within `JSSEngine`](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/ssl/javax/JSSEngine.java#L345-L355) so we can use a simple `null` check.

Reply:
In the process of doing that downcast from `SSLParameters` to `JSSParameters`, we take the ALPN value as-is (see 
https://github.com/dogtagpki/jss/pull/526/files#r503618760).  So if we construct a `JSSEngine` with an "native" `SSLParameters`, and have not overridden the ALPN data, then we get an empty list. This is why the check is necessary.
This is not a theoretical issue - I hit it with my test program.

But even so, it is just a good defensive coding practice.  If the type admits an invalid datum (it does), we should code up checks against it - or change the type (not practical here because `String[]` is imposed).

Reply:
@cipherboy I've tested this PR and it's almost there.  We really do need this length check though.
Feel free to just apply this patch or modify the relevant commit: whichever you prefer:

```patch
commit 5e09241e923b426733c444ccb1ed9295726c3080
Author: Fraser Tweedale <ftweedal@redhat.com>
Date:   Tue Oct 13 12:45:46 2020 +1100

    ALPN: guard against empty array of protocol IDs
    
    javax.net.ssl.SSLParameters sets default ALPN protocol IDs to empty
    array.  When downcasting to JSSParameters, we take the value as-is,
    so we could also end up with an empty array.  Therefore when
    deciding whether to create the ALPN extension the non-null check is
    not sufficient.  Also check that the array is non-empty.

diff --git a/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java b/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java
index 93fe703c..f84ae800 100644
--- a/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java
+++ b/org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java
@@ -347,7 +347,7 @@ public class JSSEngineReferenceImpl extends JSSEngine {
             throw new SSLException("Unable to enable SSL Handshake Callback on this SSLFDProxy instance.");
         }
 
-        if (alpn_protocols != null) {
+        if (alpn_protocols != null && alpn_protocols.length > 0) {
             byte[] wire_data = getALPNWireData();
             if (wire_data == null) {
                 throw new RuntimeException("JSSEngine.init(): ALPN wire data is NULL but alpn_protocols is non-NULL.");
```